### ISSUE TITLE
[New Feature Enhancement] Generalize DPO with different f-divergeneces

### ISSUE THREAD

**Issue by 1485840691** [huggingface/trl#1259]
### Why

There is a [paper ](https://arxiv.org/pdf/2309.16240.pdf. )discussing generalizing DPO with different f-divergences (present implementation using log() is one of its kind: reverse kL w/ a = 0) to help model better balance alignment performance and generation diversity.

According to the paper, 
"
 Empirically, adopting these f-divergences ensures a balance
 between alignment performance and generation diversity. Importantly, f-DPO
 outperforms PPO-based methods in divergence efficiency, and divergence con
straints directly influence expected calibration error (ECE).
"

### How
Would like to work out a PR to add these f-divergences besides the current supported (reverse KL)

![image](https://github.com/huggingface/trl/assets/110707330/4950eb0f-bca5-463b-829f-5dd871a60077)

![image](https://github.com/huggingface/trl/assets/140311653/75898d49-08b7-4aa5-8885-e3eeb46fe1aa)

Implementation should be straightforward.
A simple code update for illustration purposes
`
      
       rejected_rewards = policy_chosen_logps.to(self.accelerator.device) - reference_chosen_logps.to(self.accelerator.device)
       rejected_rewards_exp = torch.exp(rejected_rewards)

       chosen_rewards = policy_chosen_logps.to(self.accelerator.device) - reference_chosen_logps.to(self.accelerator.device)
        chosen_rewards_exp = torch.exp(chosen_rewards)

       # Forward KL
       logits = -1/chosen_rewards_exp - (-1 / rejected_rewards_exp)

       # Js-divergence
       logits = (chosen_rewards - torch.log(1 + chosen_rewards_exp)) - (rejected_rewards - torch.log(1 + rejected_rewards_exp))

       # alpha-divergence
       logits = (1 - chosen_rewards_exp ** (-self.alpha_div)) / self.alpha_div -  (1 - rejected_rewards_exp ** (-self.alpha_div)) / self.alpha_div `

Possible updates to existing class
`
class DPOTrainer(Trainer):
def __init__(
              ...
              **f_divergence_kwargs: Optional[Dict] = None,**
              ...
)`

Any concerns, please let me know.

**Comment by younesbelkada** [comment]
cc @kashif what do you think about this feature?

**Comment by kashif** [comment]
yeah its nice I believe, @1485840691 do you want to try to make a PR with some tests?

**Comment by 1485840691** [comment]
@kashif  yes, I would like to create a PR and plan to submit 1st version next week. After the PR is created, I plan to use the current available data to run DPO in different divergence functions, check the metrics as presented in paper and validate if anything is wrong. The experiment could be conducted while PR is reviewed at the same time.
